prusti-viper/src/encoder/encoder.rs

```diff
@@ -249,10 +259,248 @@ impl<'v, 'tcx> Encoder<'v, 'tcx> {
                 type_vars: vec![],
             });
         }
+
+        if config::enable_purification_optimization() {
+            domains.push(self.axiomatized_function_domain.borrow().clone());
+            domains.push(self.get_nat_domain());
+            domains.push(self.get_primitive_valid_domain())
+        }
+
         domains.sort_by_key(|d| d.get_identifier());
         domains
     }
 
+    fn get_primitive_valid_domain(&self) -> vir::Domain {
```

Comment:
This and the `get_nat_domain` implementation might go to `prusti-viper/src/encoder/builtin_encoder.rs` (e.g. adding a `BuiltinDomainKind`) to avoid making `encoder.rs` longer.

Reply:
As you suggested I have moved this to `builtin_encoder.rs` and added `BuiltinDomainKind`